fdbserver/MasterProxyServer.actor.cpp

```diff
@@ -82,30 +82,49 @@ Future<Void> forwardValue(Promise<T> out, Future<T> in)
 
 int getBytes(Promise<Version> const& r) { return 0; }
 
-ACTOR Future<Void> getRate(UID myID, MasterInterface master, int64_t* inTransactionCount, double* outTransactionRate) {
-	state Future<Void> nextRequestTimer = Void();
+ACTOR Future<Void> getRate(UID myID, MasterInterface master, int64_t* inTransactionCount, double* outTransactionRate, HealthMetrics* healthMetrics,
+						   GetDetailedHealthMetricsReply* getDetailedHealthMetricsReply) {
+	state Future<Void> nextRequestTimer = Never();
+	state Future<Void> nextDetailedRequestTimer = Void();
 	state Future<Void> leaseTimeout = Never();
+	state Future<Void> detailedLeaseTimeout = Never();
 	state Future<GetRateInfoReply> reply;
 	state int64_t lastTC = 0;
 
 	loop choose{
 		when(wait(nextRequestTimer)) {
 			nextRequestTimer = Never();
-			reply = brokenPromiseToNever(master.getRateInfo.getReply(GetRateInfoRequest(myID, *inTransactionCount)));
+			reply = brokenPromiseToNever(master.getRateInfo.getReply(GetRateInfoRequest(myID, *inTransactionCount, false)));
+		}
+		when(wait(nextDetailedRequestTimer)) {
+			nextDetailedRequestTimer = Never();
+			nextRequestTimer = Never();
+			reply = brokenPromiseToNever(master.getRateInfo.getReply(GetRateInfoRequest(myID, *inTransactionCount, true)));
 		}
 		when(GetRateInfoReply rep = wait(reply)) {
 			reply = Never();
+			healthMetrics->update(rep.healthMetrics, rep.detailed, true);
 			*outTransactionRate = rep.transactionRate;
 			//TraceEvent("MasterProxyRate", myID).detail("Rate", rep.transactionRate).detail("Lease", rep.leaseDuration).detail("ReleasedTransactions", *inTransactionCount - lastTC);
 			lastTC = *inTransactionCount;
 			leaseTimeout = delay(rep.leaseDuration);
 			nextRequestTimer = delayJittered(rep.leaseDuration / 2);
+			if (rep.detailed) {
```

Comment:
The reply does not need to tell us it is detailed, we should know it is detailed because the request we sent asked for details